pkg/auth/requests/authenticate.go

```diff
@@ -160,9 +176,69 @@ func (a *tokenAuthenticator) Authenticate(req *http.Request) (*AuthenticatorResp
 	authResp.Extras = getUserExtraInfo(token, u, attribs)
 	logrus.Debugf("Extras returned %v", authResp.Extras)
 
+	logrus.Debugf("Token %v, lastUsedAt: started update of `lastused`", token.ObjectMeta.Name)
+
+	// While the fix in the previous commit should have made this impossible maybe better to keep a guard.
+	if a.tokenWClient == nil {
+		logrus.Errorf("Token %v, lastUsedAt: skipping update, no wrangler client", token.ObjectMeta.Name)
+		return authResp, nil
+	}
+
+	now := time.Now().Truncate(lastUsedAtGranularity)
+	logrus.Debugf("Token %v, lastUsedAt: now is       %v", token.ObjectMeta.Name, now)
+
+	if token.LastUsedAt != nil {
+		lastRecorded := token.LastUsedAt.Time.Truncate(lastUsedAtGranularity)
+		logrus.Debugf("Token %v, lastUsedAt: recorded %v", token.ObjectMeta.Name, lastRecorded)
+
+		// throttle ... skip update if the recorded/known last use is not
+		// strictly in the past, relative to us. IOW if the token is already
+		// at the minute we want, or even ahead, then we have nothing to do.
+
+		if now.Before(lastRecorded) || now.Equal(lastRecorded) {
+			logrus.Debugf("Token %v, lastUsedAt: now <= recorded, skipped update",
+				token.ObjectMeta.Name)
+			return authResp, nil
+		}
+	}
+
+	// green light for patch
+
+	lastUsed := metav1.NewTime(now)
+	patch, err := makeLastUsedPatch(lastUsed)
+
+	if err != nil {
+		logrus.Errorf("Token %v, lastUsedAt: patch creation failed: %v", token.ObjectMeta.Name, err)
+		return nil, err
+	}
+
+	logrus.Debugf("Token %v, lastUsedAt: created patch `%v`", token.ObjectMeta.Name, string(patch))
+
+	_, err = a.tokenWClient.Patch(token.ObjectMeta.Name, types.JSONPatchType, patch)
+	if err != nil {
+		logrus.Errorf("Token %v, lastUsedAt: patch application failed: %v", token.ObjectMeta.Name, err)
+		return nil, err
```

Comment:
I'm not sure there's much value in returning the error here.

This could cause a retry of the _API Operation_ that we are recording the last use of, after we know that the API operation succeeded.

I'd be inclined to log as a warning, and not return the error.

Reply:
Fair point.

Reply:
Should we do this for line 211 too ? I.e. `patch creation failed`. Should actually not happen, not really (json marshalling).

Reply:
I just don't think there's value in returning an error case in the event of a failure to patch the last used time for the token **after** the operation completed.

Let's say this gets reported back to the user, unless we're _very careful_ they'd see the message about the update, but might not realise that the thing they wanted worked which would be really confusing?

Being careful likely results in ignoring this error (which implies that we'd do best to warn and drop) ?

Reply:
Addressed with commit 527b9883773520b17e5d7f4b5d2dc129b7f92b2c